core/envelope-sim/src/main/kotlin/fr/sncf/osrd/envelope_sim/etcs/ETCSBrakingCurves.kt

```diff
@@ -0,0 +1,221 @@
+    for (i in refCurve.pointCount() - 1 downTo 0) {
+        val speed = speeds[i]
+        val sbiPosition = getSbiPosition(pos[i], speed, tBs)
+        val permittedSpeedPosition =
+            getPermittedSpeedPosition(sbiPosition, speed, guiCurve.interpolatePosition(speed))
+        val indicationPosition = getIndicationPosition(permittedSpeedPosition, speed, tBs)
+        if (indicationPosition >= 0) {
+            reversedNewPos.add(indicationPosition)
+            reversedNewSpeeds.add(speed)
+        } else if (i != refCurve.pointCount() - 1 && reversedNewPos[i + 1] > 0) {
+            val prevPos = reversedNewPos[i + 1]
+            val prevSpeed = reversedNewSpeeds[i + 1]
+            val speedAtZero =
+                prevSpeed - prevPos * (speed - prevSpeed) / (indicationPosition - prevPos)
+            reversedNewPos.add(0.0)
+            reversedNewSpeeds.add(speedAtZero)
+            break
+        }
+    }
+
+    val nbPoints = reversedNewPos.size
+    val newPosArray = DoubleArray(nbPoints)
+    val newSpeedsArray = DoubleArray(nbPoints)
+    for (i in newPosArray.indices) {
+        newPosArray[i] = reversedNewPos[nbPoints - 1 - i]
+        newSpeedsArray[i] = reversedNewSpeeds[nbPoints - 1 - i]
+    }
+    val brakingCurve =
+        EnvelopePart.generateTimes(listOf(EnvelopeProfile.BRAKING), newPosArray, newSpeedsArray)
+
+    return brakingCurve
+}
+
+/**
+ * Keep the part of the full braking curve which is located underneath the overlay and intersects
+ * with it.
+ */
+private fun keepBrakingCurveUnderOverlay(
+    fullBrakingCurve: EnvelopePart,
+    overlay: Envelope
+): EnvelopePart {
+    assert(fullBrakingCurve.maxSpeed >= overlay.minSpeed)
+    val positions = fullBrakingCurve.clonePositions()
+    val speeds = fullBrakingCurve.cloneSpeeds()
+    val timeDeltas = fullBrakingCurve.cloneTimes()
+    val nbPoints = fullBrakingCurve.pointCount()
+    val partBuilder = EnvelopePartBuilder()
+    partBuilder.setAttr(EnvelopeProfile.BRAKING)
+    val overlayBuilder =
+        ConstrainedEnvelopePartBuilder(
+            partBuilder,
+            SpeedConstraint(0.0, EnvelopePartConstraintType.FLOOR),
+            EnvelopeConstraint(overlay, EnvelopePartConstraintType.CEILING)
+        )
+    overlayBuilder.initEnvelopePart(positions[nbPoints - 1], speeds[nbPoints - 1], -1.0)
+    for (i in fullBrakingCurve.pointCount() - 2 downTo 0) {
+        if (!overlayBuilder.addStep(positions[i], speeds[i], timeDeltas[i])) break
+    }
+    return partBuilder.build()
+}
+
+fun getSbiPosition(ebiOrSbdPosition: Double, speed: Double, tbs: Double): Double {
+    return getPreviousPosition(ebiOrSbdPosition, speed, tbs)
+}
+
+fun getPermittedSpeedPosition(
+    sbiPosition: Double,
+    speed: Double,
+    guiPosition: Double = Double.POSITIVE_INFINITY
+): Double {
+    return min(getPreviousPosition(sbiPosition, speed, tDriver), guiPosition)
+}
+
+fun getIndicationPosition(psPosition: Double, speed: Double, tBs: Double): Double {
```

Comment:
I'd name it more explicitly to "enforce" the link:
```suggestion
fun getIndicationPosition(permittedSpeedPosition: Double, speed: Double, tBs: Double): Double {
```

Reply:
Fixed in https://github.com/OpenRailAssociation/osrd/pull/10014/commits/c1bc4feb0ae2e7792d1eb8b8c9171b225d546229